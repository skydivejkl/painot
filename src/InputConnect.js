import React from "react";
import {connectLean} from "lean-redux";
import {omit, get} from "lodash/fp";

const omitConnectProps = omit(["serialize", "deserialize"]);

let InputConnect = ({component, ...props}) => {
    const Component = component;

    return <Component {...omitConnectProps(props)} />;
};
InputConnect = connectLean({
    mapState(state, props) {
        var value = state[props.name] || "";

        value = typeof props.deserialize === "function"
            ? props.deserialize(value)
            : value;

        return {value};
    },

    onChange(e) {
        const value = typeof this.props.serialize === "function"
            ? this.props.serialize(e)
            : e.target.value;

        this.setState({
            [this.props.name]: value,
        });
    },
})(InputConnect);

const defaultValidate = v => String(v || "").trim() !== "";

export const createValidator = (scope, keys, validate = defaultValidate) =>
    state => keys.every(key => validate(get([scope, key], state)));

export default InputConnect;
