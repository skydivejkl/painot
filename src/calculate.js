const JET_A = 0.804; // JET A tiheys per litra
const MTOW = 1338; // CHQ maksimi lentoonlähtö paino (kg)
const PLANE = 831; // CHQ kuivapaino (kg)

export default function calculate(
    {
        pilotti = 0,
        polttoaine = 0,
        hyppaaja1 = 0,
        hyppaaja2 = 0,
        hyppaaja3 = 0,
        hyppaaja4 = 0,
    },
) {
    var cargo = polttoaine * JET_A +
        pilotti +
        hyppaaja1 +
        hyppaaja2 +
        hyppaaja3 +
        hyppaaja4;

    var cargoMoment = polttoaine * JET_A +
        pilotti * 0.932 +
        hyppaaja1 * 0.949 +
        hyppaaja2 * 1.375 +
        hyppaaja3 * 1.507 +
        hyppaaja4 * 1.957;

    return {
        cargo: cargo,
        spare: MTOW - (PLANE + cargo),
        spareMoment: MTOW - (PLANE + cargoMoment),
    };
}
