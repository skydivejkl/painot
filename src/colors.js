export const shade0 = "#657A99";
export const shade1 = "#A5B1C1";
export const shade2 = "#8191A9";
export const shade3 = "#4B6388";
export const shade4 = "#304C75";
