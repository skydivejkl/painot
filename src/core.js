import React from "react";
import simple, {View} from "react-simple";
import {withRouter} from "react-router-dom";
import {compose, mapProps} from "recompose";

export const Text = simple(View, {
    fontFamily: "helvetica",
});

class AutoSelect extends React.Component {
    constructor(props) {
        super(props);
        this.select = this.select.bind(this);
    }

    componentDidMount() {
        this.select();
    }

    select() {
        if (!this.props.autoFocus) {
            return;
        }

        if (!this.input) {
            return;
        }

        this.input.focus();

        if (typeof this.input.select === "function") {
            this.input.select();
        }

        if (typeof this.input.setSelectionRange === "function") {
            try {
                this.input.setSelectionRange(0, 999);
            } catch (error) {
                // console.error(error);
            }
        }
    }

    render() {
        return (
            <input
                ref={input => {
                    this.input = input;
                }}
                onFocus={this.select}
                {...this.props}
            />
        );
    }
}

export const Button = simple(Text.create("button"), {
    backgroundColor: "#DDD",
    color: "black",
    fontSize: 20,
    alignItems: "center",
    justifyContent: "center",
    textDecoration: "none",
});

export const Input = simple(Text.create(AutoSelect), {
    backgroundColor: "#DDD",
    padding: 10,
    textAlign: "center",
    outline: "none",
    width: "100%",
    fontSize: 50,
});

export const Sep = simple(View, {
    width: 10,
    height: 10,
});

export const Scroll = simple(View, {
    paddingLeft: 20,
    paddingRight: 20,
    flex: 1,
    overflowX: "auto",
});

export const Title = simple(Text, {
    fontSize: 35,
    height: 100,
    overflow: "hidden",
    textAlign: "center",
    fontWeight: "bold",
    alignItems: "center",
    justifyContent: "center",
    padding: 20,
});

export const fromRoot = s => s ? "/" + s : null;

export const withRouterProps = mapper => compose(
    mapProps(props => ({originalProps: props})),
    withRouter,
    mapProps(({originalProps, ...router}) => {
        return {
            ...originalProps,
            ...mapper(router),
        };
    }),
);

export const addResultsFlag = withRouterProps(router => ({
    hasResultsFlag: router.location.search === "?results=1",
}));
