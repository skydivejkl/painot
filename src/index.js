import React from "react"; // must be in scope when using JSX
import ReactDOM from "react-dom";

console.log("git date", GIT_COMMIT_DATE);
console.log("git message", GIT_COMMIT_MESSAGE);
console.log("git rev", GIT_COMMIT_REV);

function HelloWorld() {
    return (
        <div>React.js Hello World!</div>
    );
}


const container = document.getElementById("app-container");
ReactDOM.render(<HelloWorld />, container);
