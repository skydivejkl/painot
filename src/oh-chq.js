export const JET_A = 0.804; // JET A tiheys
export const MTOW = 1338; // CHQ suurin lentoonlähtöpaino (kg)

// OH-CHQ punnitustodistuksesta 25.3.2015
export const PLANE = 838.98; // CHQ kuivapaino (kg)
export const PLANE_MOMENT = 782.87; // CHQ perusmomentti

export const items = [
    {
        name: "fuel",
        title: "JET A",
        description: "Tankattu JET A litroissa",
        position: 1.22,
        convert(value) {
            return parseFloat(value, 10) * JET_A;
        },
    },
    {
        name: "pilot",
        title: "pilotti",
        description: "Pilotin paino pelastusvarjoineen (kg)",
        position: 0.932,
    },
    {
        name: "jumper1",
        title: '"mesun paikalla"',
        description: 'ns. "mesun paikalla" istuvan hyppääjän exit paino jolla on selkä menosuuntaan. (kg)',
        position: 0.949,
    },
    {
        name: "jumper2",
        title: "ovella",
        description: "Oven vieressä istuva hyppääjä exit paino jolla on naama menosuuntaan (kg)",
        position: 1.375,
    },
    {
        name: "jumper3",
        title: "pilotin penkin takana",
        description: "Pilotin penkin takana istuva hyppääjän exit paino (kg)",
        position: 1.507,
    },
    {
        name: "jumper4",
        title: "perällä",
        description: "Ihan perällä istuva hyppääjän exit paino (kg)",
        position: 1.957,
    },
];

// Taulukon arvot
// https://raw.githubusercontent.com/skydivejkl/painot/master/img/sma-gc-limits.png
const MAX_GC = 1.17;
const MAX_MASS = 1338;
const SLOPE_START = 1040;
const SLOPE = 0.08;
const MIN_GC = 0.92;

export function isGCWithinLimits(mass, gc) {
    // suurin gc taulukossa
    if (gc > MAX_GC) {
        return false;
    }

    // suurin paino taulukossa
    if (mass > MAX_MASS) {
        return false;
    }

    if (mass > SLOPE_START && mass <= MAX_MASS) {
        const bottomLimit = MIN_GC +
            SLOPE * (mass - SLOPE_START) / (MAX_MASS - SLOPE_START);
        return gc > bottomLimit;
    }

    if (mass <= SLOPE_START) {
        return gc > MIN_GC;
    }

    console.error("Failed to calculate GC limit for", mass, gc);
    return false;
}
