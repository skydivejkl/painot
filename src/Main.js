import React from "react";
import {connect} from "react-redux";
import {Link, Redirect, withRouter} from "react-router-dom";
import simple, {View, css} from "react-simple";
import {Route} from "react-router-dom";
import {get, omit, last, mapValues} from "lodash/fp";
import {compose, withHandlers, mapProps} from "recompose";

import Touchable from "./Touchable";
import TopNav from "./TopNav";
import InputConnect from "./InputConnect";
import {Text, Input, Sep, Button} from "./core";
import calculate from "./calculate";

css.global("body, html", {
    padding: 0,
    margin: 0,
    backgroundColor: "black",
});

const Container = simple(View, {
    position: "absolute",
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    alignItems: "center",
});

const Wrap = simple(View, {
    backgroundColor: "white",
    flex: 1,
    width: "100%",
    // overflowX: "auto",
    "@media (min-width: 450px)": {
        width: 450,
    },
});

const Title = simple(Text, {
    fontSize: 35,
    height: 100,
    overflow: "hidden",
    textAlign: "center",
    fontWeight: "bold",
    alignItems: "center",
    justifyContent: "center",
    padding: 20,
});

const Description = simple(Text, {
    textAlign: "center",
    color: "black",
    fontSize: 12,
});

var NumInput = props => (
    <InputConnect
        component={Input}
        type="number"
        pattern="[0-9]*"
        inputMode="numeric"
        scope="painot"
        {...props}
    />
);

const omitRouterProps = omit([
    "nextNavigate",
    "backNavigate",
    "length",
    "location",
    "createHref",
    "push",
    "replace",
    "go",
    "goBack",
    "goForward",
    "block",
    "listen",
    "match",
]);
NumInput = compose(
    withRouter,
    withHandlers({
        onKeyDown: props => e => {
            if (e.key === "Enter" || e.key === "Tab") {
                e.preventDefault();
                props.push(props.nextNavigate);
            }

            if (e.shiftKey && e.key === "Tab") {
                e.preventDefault();
                props.push(props.backNavigate);
            }
        },
    }),
    mapProps(omitRouterProps),
)(NumInput);

const inputs = [
    {
        name: "fuel",
        title: "Polttoaine",
        description: "Tankattu JET A litroissa",
    },
    {name: "pilot", title: "Pilotti", description: "Lentäjän paino (kg)"},
    {
        name: "jumper1",
        title: '"Mesun paikka"',
        description: 'ns. "mesun paikalla" istuvan hyppääjän exit paino jolla on selkä menosuuntaan. (kg)',
    },
    {
        name: "jumper2",
        title: "Ovella",
        description: "Oven vieressä istuva hyppääjä jolla on naama menosuuntaan (kg)",
    },
    {
        name: "jumper3",
        title: "Pilotin penkin takana",
        description: "Lentäjän penkin takana istuva hyppääjän exit paino (kg)",
    },
    {
        name: "jumper4",
        title: "Perällä",
        next: "/tulos",
        nextText: "Tulokset",
        description: "Ihan perällä istuva hyppääjän exit paino (kg)",
    },
];

const Flex = simple(View, {
    flex: 1,
});

const Form = ({back, next, nextText, title, name, description}) => (
    <Flex>
        <TopNav back={back} nextText={nextText} next={next} />
        <Scroll>

            <Title>{title}</Title>

            <NumInput
                autoFocus
                name={name}
                placeholder="0"
                nextNavigate={next}
                backNavigate={back}
            />

            <Sep />

            <Description>
                {description}
            </Description>
        </Scroll>
    </Flex>
);

const TotalText = simple(Text, {
    width: 150,
    fontSize: 19,
});

const ResultText = simple(
    Text,
    {
        fontSize: 25,
        width: 150,
        fontWeight: "bold",
        color: "#3cd03c",
    },
    {
        bad: {
            color: "red",
        },
    },
);

const mapObValuesToFloats = mapValues(val => parseFloat(val, 10));

const Row = simple(View, {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
});

const Img = simple(View.create("img"), {
    margin: "auto",
    marginTop: 20,
    width: "100%",
});

const Scroll = simple(View, {
    paddingLeft: 20,
    paddingRight: 20,
    flex: 1,
    overflowX: "auto",
});

const Zoom = simple(Button.create("a"), {
    padding: 15,
    margin: 10,
    width: "50%",
    fontSize: 15,
    alignSelf: "center",
});

var Results = ({cargo, spare, total, gc, gcOk, mtow}) => (
    <Flex>
        <TopNav back={fromRoot(last(inputs).name)} next="/" nextText="Alkuun" />
        <Scroll>

            <Title>Tulokset</Title>

            <Row>
                <TotalText>Kuorma</TotalText>
                <ResultText bad={total > mtow}>
                    {cargo.toFixed(2)} kg
                </ResultText>
            </Row>

            <Sep />

            <Row>
                <TotalText>Kokonaispaino</TotalText>
                <ResultText bad={mtow < total}>
                    {total.toFixed(2)} kg
                </ResultText>
            </Row>

            <Sep />

            <Row>
                <TotalText>Vapaana</TotalText>
                <ResultText bad={spare < 0}>{spare.toFixed(2)} kg</ResultText>
            </Row>

            <Sep />

            <Row>
                <TotalText>Massakeskipiste</TotalText>
                <ResultText bad={!gcOk}>{gc.toFixed(3)} m</ResultText>
            </Row>

            <Img
                src="https://raw.githubusercontent.com/skydivejkl/painot/master/img/sma-gc-limits.png"
            />

            <Zoom
                href="https://raw.githubusercontent.com/skydivejkl/painot/master/img/sma-gc-limits.png"
            >
                ZOOM
            </Zoom>
        </Scroll>
    </Flex>
);
Results = connect(state => calculate(mapObValuesToFloats(state.painot)))(
    Results,
);

const fromRoot = s => s ? "/" + s : null;

const Main = () => (
    <Container>
        <Wrap>

            <Route
                exact
                path="/"
                render={() => <Redirect to={"/" + inputs[0].name} />}
            />

            {inputs.map((item, index, array) => {
                const next = item.next ||
                    fromRoot(get([index + 1, "name"], array));
                const back = fromRoot(get([index - 1, "name"], array));
                return (
                    <Route
                        key={item.name}
                        path={"/" + item.name}
                        render={() => (
                            <Form {...item} next={next} back={back} />
                        )}
                    />
                );
            })}

            <Route path="/tulos" component={Results} />
        </Wrap>
    </Container>
);

export default Main;
