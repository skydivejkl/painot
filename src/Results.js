import React from "react";
import {connect} from "react-redux";
import simple, {View} from "react-simple";
import {last, mapValues} from "lodash/fp";

import {Text, Button, Scroll, Title, Sep, fromRoot} from "./core";
import TopNav from "./TopNav";
import calculate, {dataInputs} from "./calculate";

const mapObValuesToFloats = mapValues(val => parseFloat(val, 10));

const Row = simple(View, {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
});

const Flex = simple(View, {
    flex: 1,
});

const Img = simple(View.create("img"), {
    margin: "auto",
    marginTop: 20,
    width: "100%",
});

const Zoom = simple(Button.create("a"), {
    padding: 15,
    margin: 10,
    width: "50%",
    fontSize: 15,
    alignSelf: "center",
});

const TotalText = simple(Text, {
    width: 150,
    fontSize: 19,
});

const ResultText = simple(
    Text,
    {
        fontSize: 25,
        width: 150,
        fontWeight: "bold",
        color: "#3cd03c",
    },
    {
        bad: {
            color: "red",
        },
    },
);

export const Results = ({cargo, spare, total, gc, gcOk, mtow}) => (
    <Flex>
        <TopNav
            back={fromRoot(last(dataInputs).name)}
            next="/"
            nextText="Alkuun"
        />
        <Scroll>

            <Title>Tulokset</Title>

            <Row>
                <TotalText>Kuorma</TotalText>
                <ResultText bad={total > mtow}>
                    {cargo.toFixed(2)} kg
                </ResultText>
            </Row>

            <Sep />

            <Row>
                <TotalText>Kokonaispaino</TotalText>
                <ResultText bad={mtow < total}>
                    {total.toFixed(2)} kg
                </ResultText>
            </Row>

            <Sep />

            <Row>
                <TotalText>Vapaana</TotalText>
                <ResultText bad={spare < 0}>{spare.toFixed(2)} kg</ResultText>
            </Row>

            <Sep />

            <Row>
                <TotalText>Massakeskipiste</TotalText>
                <ResultText bad={!gcOk}>{gc.toFixed(3)} m</ResultText>
            </Row>

            <Img
                src="https://raw.githubusercontent.com/skydivejkl/painot/master/img/sma-gc-limits.png"
            />

            <Zoom
                href="https://raw.githubusercontent.com/skydivejkl/painot/master/img/sma-gc-limits.png"
            >
                ZOOM
            </Zoom>
        </Scroll>
    </Flex>
);

export default connect(state => calculate(mapObValuesToFloats(state.painot)))(
    Results,
);
