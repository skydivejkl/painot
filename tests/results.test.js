import React from "react";
import renderer from "react-test-renderer";
import {MemoryRouter} from "react-router";

import {Results} from "../src/Results";
import calculate from "../src/calculate";

test("render results", () => {
    const data = {
        fuel: 50,
        pilot: 80,
        jumper1: 105,
        jumper2: 90,
        jumper3: 85,
        jumper4: 85,
    };

    const tree = renderer
        .create(
            <MemoryRouter>
                <Results {...calculate(data)} />
            </MemoryRouter>,
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
