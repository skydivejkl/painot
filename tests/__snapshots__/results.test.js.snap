exports[`test render results 1`] = `
<div
  className="Flex___0 css-1mobb16">
  <div
    className="Container___1 css-1a1kkk0">
    <div
      className="View___3 FlexTouchable___2 css-mfy564 css-989zum"
      onTouchEnd={[Function]}
      onTouchStart={[Function]}>
      <a
        className="FlexLink___4 css-gox86g"
        href="/jumper4"
        onClick={[Function]}>
        Edellinen
      </a>
    </div>
    <div
      className="Sep___5 css-3baico" />
    <div
      className="View___7 FlexTouchable___6 css-mfy564 css-989zum"
      onTouchEnd={[Function]}
      onTouchStart={[Function]}>
      <a
        className="FlexLink___8 css-gox86g"
        href="/"
        onClick={[Function]}>
        Alkuun
      </a>
    </div>
  </div>
  <div
    className="Scroll___9 css-1az1c6z">
    <div
      className="Title___10 css-1lc423o">
      Tulokset
    </div>
    <div
      className="Row___11 css-1v992oz">
      <div
        className="TotalText___12 css-1afjqs1">
        Kuorma
      </div>
      <div
        className="ResultText___13 css-1ackb39">
        485.20
         kg
      </div>
    </div>
    <div
      className="Sep___14 css-3baico" />
    <div
      className="Row___15 css-1v992oz">
      <div
        className="TotalText___16 css-1afjqs1">
        Kokonaispaino
      </div>
      <div
        className="ResultText___17 css-1ackb39">
        1324.18
         kg
      </div>
    </div>
    <div
      className="Sep___18 css-3baico" />
    <div
      className="Row___19 css-1v992oz">
      <div
        className="TotalText___20 css-1afjqs1">
        Vapaana
      </div>
      <div
        className="ResultText___21 css-1ackb39">
        13.82
         kg
      </div>
    </div>
    <div
      className="Sep___22 css-3baico" />
    <div
      className="Row___23 css-1v992oz">
      <div
        className="TotalText___24 css-1afjqs1">
        Massakeskipiste
      </div>
      <div
        className="ResultText___25 css-1ackb39">
        1.076
         m
      </div>
    </div>
    <img
      className="Img___26 css-5bhlr8"
      src="https://raw.githubusercontent.com/skydivejkl/painot/master/img/sma-gc-limits.png" />
    <a
      className="Zoom___27 css-vpopo7"
      href="https://raw.githubusercontent.com/skydivejkl/painot/master/img/sma-gc-limits.png">
      ZOOM
    </a>
  </div>
</div>
`;
